### PR TITLE
Add linearRelu op to linalg structured ops

### PR DESCRIPTION
**Pull request by flemairen6**

### CODE REVIEW

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml [hunk 1]
```diff
@@ -4096,38 +4096,39 @@ structured_op: !LinalgStructuredOpConfig
     name: I
     kind: input_tensor
     type_var: T1
-    shape_map: affine_map<()[s0, s1, s2, s3, s4, s5, s6, s7, s8, s9] -> (s0, s9, s1 *
-      s2 + s3 * s4, s5 * s6 + s7 * s8)>
+    shape_map: affine_map<()[s0, s1, s2, s3, s4, s5, s6, s7, s8, s9] -> (s0, s1, s2
```
Comment (ljfitz) [inline mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml:4099]: @gargaroff - you've seen unrelated changes like these before when running the script to apply `core_named_ops.py`. Do you know why it happens?

File: mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp [hunk 2]
```diff
@@ -698,6 +699,50 @@ struct LinearLowering : OpRewritePattern<LinearOp> {
...
+        loc, outputType.getShape(), biasType.getElementType());
+    Value broadcastBiasOp =
+        rewriter.create<Broadcast1DTo2DOp>(loc, bias, broadcastedBias)
+            ->getResult(0);
+
+    auto linearResult = rewriter.create<MatmulOp>(loc,
+                  outputType, ValueRange{op.getOperand(0), transposeWeightsOp},
+                  broadcastBiasOp).getResult(0);
+
+    rewriter.replaceOpWithNewOp<Relu2DNchwOp>(
```
Comment (ljfitz) [inline mlir/lib/Dialect/Linalg/Transforms/Unfuse.cpp:737]: NIT: it is disappointing you didn't find a way to avoid copy-paste by e.g. moving the body of `LinearLowering::matchAndRewrite` into a new template function and doing a `rewriter.replaceOp()` with the result there, while and setting `linearResult` here.

**Review by ljfitz — APPROVED** [review]

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml (hunk 1 above)
Comment (gargaroff) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml:4099]: I haven't investigated this in full, but it looks to me like that this file is simply out-of-date and somebody forgot to update it.

Comment (ljfitz) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml:4099]: Thanks for the reminder. @flemairen6 - please do not bring in these unrelated changes as we don't want to be more different from upstream than is essential.

Comment (ljfitz) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml:4099]: Thanks for the reminder. @flemairen6 - please do not bring in these unrelated changes as we don't want to be more different from upstream than is essential.

**Review by ljfitz — CHANGES_REQUESTED** [review]

File: mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py [hunk 3]
```diff
@@ -1377,3 +1377,38 @@ def fill_rng_2d(min=ScalarDef(F64),
...
+
+@linalg_structured_op
+def linear_relu(
+    I=TensorDef(T1, S.W, S.H),
+    W=TensorDef(T1,  S.K, S.H),
+    B=TensorDef(T1,  S.K),
+    O=TensorDef(T1,  S.W, S.K, output=True)):
+  """Performs a linear/fully-connected + relu operation
+
+  This is a long description that I'll fill later
```
Comment (gargaroff) [inline mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py:1389]: When is later? :p

File: mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py [hunk 4]
```diff
@@ -1377,3 +1377,38 @@ def fill_rng_2d(min=ScalarDef(F64),
...
+    OFM=TensorDef(T1, Batch, S.C, output=True )):
+  """Applies the ReLU activation function to every value in the tensor.
+  
+  Layout:
+    * Input: NC
+  """
+  domain(D.b, D.c)
+  OFM[D.b, D.c] = BinaryFn.max_signed(
+    IFM[D.b, D.c], TypeFn.cast_signed(T1, const(0.0))
+  )
```
Comment (gargaroff) [inline mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py:1414]: Can you end a newline at the end of the file?

**Review by gargaroff — APPROVED** [review]
